contracts/TokenDistributor.sol

```diff
@@ -0,0 +1,212 @@
+            revert OffsetTooBig();
+        }
+        offset = daysToOffsetBy * 1 days;
+        uint startOfThisWeek = _startOfWeek(block.timestamp);
+        startTime = startOfThisWeek;
+        lastCheckpoint = startOfThisWeek;
+    }
+
+    /// @inheritdoc ITokenDistributor
+    function checkpointToken() public override {
+        uint tokenBalance = kwenta.balanceOf(address(this));
+        uint toDistribute = tokenBalance - lastTokenBalance;
+        lastTokenBalance = tokenBalance;
+
+        uint previousCheckpoint = lastCheckpoint;
+        uint sinceLast = block.timestamp - previousCheckpoint;
+        lastCheckpoint = block.timestamp;
+        uint thisWeek = _startOfWeek(previousCheckpoint);
+        uint nextWeek = 0;
+
+        /// @dev Loop for potential missed weeks
+        /// iterates until caught up, unlikely to go to 52 weeks
+        for (uint i = 0; i < WEEKS_IN_YEAR; i++) {
+            nextWeek = thisWeek + 1 weeks;
+
+            if (block.timestamp < nextWeek) {
+                /// @dev if in the current week
+                if (sinceLast == 0 && block.timestamp == previousCheckpoint) {
+                    /// @dev If no time change since last checkpoint just add new tokens
+                    /// that may have been deposited (same block)
+                    tokensPerEpoch[thisWeek] += toDistribute;
+                } else {
+                    /// @dev In the event that toDistribute contains tokens
+                    /// for multiple weeks we take the remaining portion
+                    tokensPerEpoch[thisWeek] +=
+                        (toDistribute *
+                            (block.timestamp - previousCheckpoint)) /
+                        sinceLast;
+                }
+                break;
+            } else {
+                /// @dev If passed weeks missed
+                if (sinceLast == 0 && nextWeek == previousCheckpoint) {
+                    tokensPerEpoch[thisWeek] += toDistribute;
+                } else {
+                    /// @dev Store proportion of tokens for this week in the past
+                    tokensPerEpoch[thisWeek] +=
+                        (toDistribute * (nextWeek - previousCheckpoint)) /
+                        sinceLast;
+                }
+            }
+            previousCheckpoint = nextWeek;
+            thisWeek = nextWeek;
+        }
+        emit CheckpointToken(block.timestamp, toDistribute);
+    }
+
+    /// @inheritdoc ITokenDistributor
+    function claimEpoch(address to, uint epochNumber) public override {
+        _checkpointWhenReady();
+        _isEpochReady(epochNumber);
+        if (claimedEpochs[to][epochNumber] == true) {
+            revert CannotClaimTwice();
+        }
+        claimedEpochs[to][epochNumber] = true;
+
+        uint256 totalStaked = stakingRewardsV2.totalSupplyAtTime(
+            _startOfEpoch(epochNumber)
+        );
+        if (totalStaked == 0) {
+            revert NothingStakedThatEpoch();
```

Comment:
**Issue:** You also query `totalSupplyAtTime` in `calculateEpochFees` and `claimEpoch`.

This check is unnecessary as you have the following check in `calculateEpochFees`:
```solidity
        if (totalStaked == 0) {
            return 0;
        }
```

Hence you don't need the check or error `NothingStakedThatEpoch` at all, as `CannotClaim0Fees` will be triggered in that case.

This one is particularly important as `totalSupplyAtTime` could be very expensive as the number of checkpoints to binary search grows. So best not to call it twice if not needed.